_drafts/test-automation-renode.md

```diff
@@ -0,0 +1,886 @@
+testing! In a previous post, we wrote about [building firmware in
+CircleCI]({% post_url 2019-09-17-continuous-integration-for-firmware %}). This
+time, we are going to use Github Actions to build and test our firmware, as it's
+likely the easiest for most people to get up and running.
+
+To start, we'll want to start by creating a file `.github/workflows/main.yml` in
+our repo. The `blank` example from Github's
+[starter-workflows](https://github.com/actions/starter-workflows/blob/master/ci/blank.yml)
+gives a quick introduction to how the system works. I've copied it down below
+for easy reference.
+
+```yaml
+# This is a basic workflow to help you get started with Actions
+
+name: CI
+
+# Controls when the action will run. Triggers the workflow on push or pull request
+# events but only for the master branch
+on:
+  push:
+    branches: [master]
+  pull_request:
+    branches: [master]
+
+# A workflow run is made up of one or more jobs that can run sequentially or in parallel
+jobs:
+  # This workflow contains a single job called "build"
+  build:
+    # The type of runner that the job will run on
+    runs-on: ubuntu-latest
+
+    # Steps represent a sequence of tasks that will be executed as part of the job
+    steps:
+      # Checks-out your repository under $GITHUB_WORKSPACE, so your job can access it
+      - uses: actions/checkout@v2
+
+      # Runs a single command using the runners shell
+      - name: Run a one-line script
+        run: echo Hello, world!
+
+      # Runs a set of commands using the runners shell
+      - name: Run a multi-line script
+        run: |
+          echo Add other actions to build,
+          echo test, and deploy your project.
+```
+
+### Building Firmware in CI
+
+It's time to adapt the sample above to make it build our firmware. First, we'll
+need to make sure that the `checkout` routine clones submodules as well, since
+our example repo has a couple of them.
+
+```yaml
+jobs:
+  build_and_test:
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v2
+        with: #
+          submodules: true # CHANGED
+```
+
+Next, we'll want to download and configure the ARM embedded toolchain since it
+isn't installed by default. Thankfully, there is already an Action,
+[`fiam/arm-none-eabi-gcc`](https://github.com/fiam/arm-none-eabi-gcc), that we
+can use to install the toolchain for us.
+
+```yaml
+    steps:
+    [...]
+
+    # Get the arm-non-eabi-gcc toolchain
+    - name: Install arm-none-eabi-gcc
+      uses: fiam/arm-none-eabi-gcc@v1
+      with:
+          release: '9-2019-q4'
+```
+
+The last thing we need to do is to invoke `make` itself to start the build.
+
+```yaml
+    steps:
+    [...]
+
+    - name: Build Firmware
+      run: make -j4
+```
+
+### Running Renode in CI
+
+The next thing we need to do is to get Renode working in CI. For the example,
+this is as simple as running `./docker-test.sh` since the Github Actions Ubuntu
```

Comment:
You don't include any additional details about `docker-test.sh`. What is it? Where does it come from? Why do I need it? Where do I find a link to the full file?